### DISCUSSION TITLE
Implement recent searches functionality within live searchbar

### DISCUSSION THREAD

**Discussion by RolandHesse** [neuefische/questions#782] [Web] [answered]
Hi there!

Here is the current state of our search bar feature/branch: https://github.com/PeetThesing/capstone-project_snippet-manager/pull/50 

There is a number of problems for which we couldn't find a solution so far:

1. On hitting the enter button, the search is canceled and the page refreshes.

2. In case the search term doesn't match any snippet, all snippets are shown instead of none.

3. Last but not least (...), the last searches array updates with every keystroke. This is obviously due to the fact that we currently have a live search (the handleSearch()-function being connected to the onChange-attribute not to the onSubmit-attribute).

Ideally, what we would probably like to have is the following: a live-search functionality where the lastSearches-array is updated only when the user either hits the enter button or clicks outside the search-bar. (However, we are not entirely sure whether this makes sense conceptually...)

PS It goes without saying that the styling of the div in which the last searches are rendered would still have to be adapted.
PPS There are some lines of code which we commented out and with the help of which we tried to achieve the desired outcome described above, alas, without success. We still left them in case they are of any use...

**Comment by NikRoe** [comment, marked as answer]
Hi there,

I'll try answering your questions one by one:

1. Hitting the enter button triggers the default behaviour of your form. Your page tries to submit it. -> You need to prevent the default. 
To solve this, you can adapt the opening tag of your form to look like this:
```js
<StyledSearchBarForm onSubmit={(event) => event.preventDefault()}>
```

2. The reason for this behaviour is this line of code:
```js
<SnippetCardList data={results.length > 0 ? results : data} />
```
Whenever there are no results, you pass `data` to your SnippetCardList and `data` represents all snippets.

Solving this depends on how you want your application to behave.
Two options that come to my mind are: 
a. Finding no results still displays all snippets but also a text, letting the users know that no matching snippet could be found and thus all snippets are displayed.
b. You use the another instance of the SnippetCardList and based on a state that gets toggled by your search, you either show the SnippetCardList with all snippets or only the search results with a message indicating that nothing could be found
c. Maybe you have an even better idea than me!

3. Having a submit button would make this a lot easier, since you could add a new entry to the history whenever the user submits. 

a. If you wish to only work with the onChange, it is also possible, but you need to find out somehow, when the search has ended. 
One idea would be to assume certain patterns, like a space, indicate the end of a search. I built a small [codesandbox as a demo](https://codesandbox.io/p/sandbox/livesearch-with-custom-history-5h8snx?file=%2Fsrc%2FApp.js%3A35%2C6).
But this is not my preferred idea. 

b. You can also think about working with a debounce. That only adds a term to the history, if there was user inactivity for a certain amount of time. I built a [demo in codesandbox](https://codesandbox.io/p/sandbox/livesearch-with-debounce-and-search-history-qhjv9d?file=%2Fsrc%2FApp.js%3A30%2C27) for this. You might want to play around with the time you pass to the setTimeout. In the example it's 500ms.

c. This one now actually answers your point 3. It could look similar to this [demo](https://codesandbox.io/p/sandbox/livesearch-with-debounce-and-search-history-forked-t4kcvt?file=%2Fsrc%2FApp.js%3A30%2C5).

I would decide between options 3b and 3c.

I hope this helps you on your way to create a fantastic search history 🙌

**Reply by RolandHesse** [reply]
Hey Niklas, thanks a lot for for your super-helpful reply and the effort you put in it! With your help, we now have all the functionality we wanted! 🤓 🙏
